File: Garv_Singh_Chouhan_java_training/session2/usermanagement/src/main/java/com/session2/usermanagement/controller/UserController.java
package com.session2.usermanagement.controller;

import com.session2.usermanagement.dto.UserRequestDTO;
import com.session2.usermanagement.model.User;
import com.session2.usermanagement.service.UserService;
import org.springframework.web.bind.annotation.*;

import java.util.List;

// Controller handles API requests
// No business logic should be written here

@RestController
@RequestMapping("/users")
public class UserController {

    private final UserService userService;

    // Constructor Injection
    public UserController(UserService userService) {
        this.userService = userService;
    }

    // GET /users
    @GetMapping
    public List<User> getAllUsers() {
        return userService.getAllUsers();
    }

    // POST /users
    @PostMapping
    public User createUser(@RequestBody UserRequestDTO dto) {
        return userService.createUser(dto);
    }

    // GET /users/{id}
    @GetMapping("/{id}")
    public User getUserById(@PathVariable Long id) {
        return userService.getUserById(id);
    }
}

File: Garv_Singh_Chouhan_java_training/session3/usersearchsystem/src/main/java/com/session3/usersearchsystem/controller/UserController.kt
package com.session3.usersearchsystem.controller

import com.session3.usersearchsystem.dto.UserRequestDTO
import com.session3.usersearchsystem.service.UserService
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.*
import com.session3.usersearchsystem.model.User

// Controller layer handles API requests
// No business logic should be written here
@RestController
@RequestMapping("/users")
class UserController // Constructor injection (MANDATORY as per assignment)
    (private val userService: UserService) {
    // SEARCH API
    // If params are null → returns all users
    @GetMapping("/search")
    fun searchUsers(
        @RequestParam(required = false) name: String?,
        @RequestParam(required = false) age: Int?,
        @RequestParam(required = false) role: String?
    ): MutableList<User?> {
        return userService.searchUsers(name, age, role)
    }

    // SUBMIT API
    @PostMapping("/submit")
    fun submitUser(@RequestBody dto: UserRequestDTO?): ResponseEntity<*> {
        try {
            val user: User? = userService.submitUser(dto)

            // returning 201 CREATED
            return ResponseEntity.status(201).body<Any?>(user)
        } catch (e: IllegalArgumentException) {
            // returning 400 BAD REQUEST

            return ResponseEntity.badRequest().body<String?>(e.message)
        }
    }

    // DELETE API with confirmation
    @DeleteMapping("/{id}")
    fun deleteUser(
        @PathVariable id: Long?,
        @RequestParam(required = false) confirm: Boolean?
    ): String {
        return userService.deleteUser(id, confirm)
    }
}